File: crates/l1-kernel-rs/src/load_adaptive.rs
//! Provider-neutral load-adaptive worker-pool control law candidate.

use serde::{Deserialize, Serialize};

/// EWMA smoothing factor mirrored from Python `params.api`.
pub const LOAD_ADAPTIVE_EWMA_ALPHA: f64 = 0.3;
/// Queue ratio below which the controller shrinks.
pub const LOAD_ADAPTIVE_LOW_RATIO: f64 = 0.2;
/// Queue ratio above which the controller grows.
pub const LOAD_ADAPTIVE_HIGH_RATIO: f64 = 0.6;
/// Consecutive out-of-band samples required before a decision.
pub const LOAD_ADAPTIVE_HYSTERESIS_SAMPLES: u32 = 3;
/// Cooldown duration after a growth or shrink decision.
pub const LOAD_ADAPTIVE_COOLDOWN_S: f64 = 5.0;
/// Number of workers added by a normal growth decision.
pub const LOAD_ADAPTIVE_GROW_STEP: u64 = 2;
/// Worker-count divisor used by a shrink decision.
pub const LOAD_ADAPTIVE_SHRINK_FACTOR: u64 = 2;
/// Fraction of the task timeout used to identify slow tasks.
pub const LOAD_ADAPTIVE_SLOW_TASK_RATIO: f64 = 0.5;
/// Default task timeout used by the slow-task threshold.
pub const WORKER_POOL_TASK_TIMEOUT: f64 = 30.0;

/// Control action emitted by the load-adaptive controller.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum Action {
    /// Keep the current worker count.
    #[default]
    Hold,
    /// Add the configured growth step.
    Grow,
    /// Reduce the worker count by the configured divisor.
    Shrink,
    /// Add twice the normal growth step for slow tasks.
    GrowFast,
}

/// Load signals consumed by one control-law cycle.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(default)]
pub struct ControllerMetrics {
    /// Current queued work divided by queue capacity.
    pub queue_ratio: f64,
    /// Completed work rate, reserved for future control policies.
    pub completion_rate: f64,
    /// Active worker ratio, reserved for future control policies.
    pub active_ratio: f64,
    /// Elapsed time of the sampled task in seconds.
    pub task_elapsed: f64,
    /// Current resident worker count.
    pub worker_count: u64,
    /// Minimum resident worker count.
    pub worker_min: u64,
    /// Maximum resident worker count.
    pub worker_max: u64,
}

impl Default for ControllerMetrics {
    fn default() -> Self {
        Self {
            queue_ratio: 0.0,
            completion_rate: 0.0,
            active_ratio: 0.0,
            task_elapsed: 0.0,
            worker_count: 0,
            worker_min: 1,
            worker_max: 32,
        }
    }
}

/// Configuration values for one controller instance.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct ControllerConfig {
    /// Lower target-band boundary.
    pub low_ratio: f64,
    /// Upper target-band boundary.
    pub high_ratio: f64,
    /// EWMA smoothing factor.
    pub ewma_alpha: f64,
    /// Number of consecutive out-of-band samples required.
    pub hysteresis_samples: u32,
    /// Cooldown duration in seconds.
    pub cooldown_s: f64,
    /// Normal worker growth step.
    pub grow_step: u64,
    /// Worker shrink divisor.
    pub shrink_factor: u64,
    /// Slow-task threshold as a fraction of the task timeout.
    pub slow_task_ratio: f64,
    /// Task timeout in seconds.
    pub task_timeout: f64,
}

impl Default for ControllerConfig {
    fn default() -> Self {
        Self {
            low_ratio: LOAD_ADAPTIVE_LOW_RATIO,
            high_ratio: LOAD_ADAPTIVE_HIGH_RATIO,
            ewma_alpha: LOAD_ADAPTIVE_EWMA_ALPHA,
            hysteresis_samples: LOAD_ADAPTIVE_HYSTERESIS_SAMPLES,
            cooldown_s: LOAD_ADAPTIVE_COOLDOWN_S,
            grow_step: LOAD_ADAPTIVE_GROW_STEP,
            shrink_factor: LOAD_ADAPTIVE_SHRINK_FACTOR,
            slow_task_ratio: LOAD_ADAPTIVE_SLOW_TASK_RATIO,
            task_timeout: WORKER_POOL_TASK_TIMEOUT,
        }
    }
}

/// Structured configuration errors at the candidate boundary.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ControllerError {
    /// The target-band boundaries are invalid.
    InvalidRatios,
    /// The EWMA factor is outside the open/closed interval `(0, 1]`.
    InvalidAlpha,
    /// A zero shrink divisor would make a shrink decision undefined.
    InvalidShrinkFactor,
}

/// One decision returned by a control-law cycle.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Decision {
    /// Chosen control action.
    pub action: Action,
    /// Worker target after the action.
    pub target_workers: u64,
    /// Smoothed queue ratio.
    pub ewma_depth: f64,
    /// Whether the cycle was suppressed by cooldown.
    pub in_cooldown: bool,
    /// Stable human-readable reason retained for observability.
    pub reason: String,
}

/// Serializable controller state for inspection without exposing internals.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct StateSnapshot {
    /// Smoothed queue ratio rounded to four decimal places.
    pub ewma: f64,
    /// Consecutive out-of-band sample count.
    pub out_of_bounds_count: u32,
    /// Elapsed time since the last decision, supplied by the caller.
    pub last_decision_elapsed: f64,
    /// Number of decisions emitted so far.
    pub decisions_total: u64,
    /// Effective lower target-band boundary.
    pub low_ratio: f64,
    /// Effective upper target-band boundary.
    pub high_ratio: f64,
    /// Effective EWMA factor.
    pub ewma_alpha: f64,
    /// Effective hysteresis sample count.
    pub hysteresis_samples: u32,
    /// Effective cooldown duration.
    pub cooldown_s: f64,
    /// Effective growth step.
    pub grow_step: u64,
    /// Effective shrink divisor.
    pub shrink_factor: u64,
}

#[derive(Debug, Clone, Copy, Default)]
struct ControllerState {
    ewma: f64,
    out_of_bounds_count: u32,
    last_decision_at: f64,
    decisions_total: u64,
}

/// Pure load-adaptive worker-pool control law.
pub struct LoadAdaptiveController {
    config: ControllerConfig,
    state: ControllerState,
}

impl LoadAdaptiveController {
    /// Create a controller after validating its target band and smoothing factor.
    pub fn new(config: ControllerConfig) -> Result<Self, ControllerError> {
        if config.low_ratio < 0.0
            || config.low_ratio >= config.high_ratio
            || config.high_ratio > 1.0
        {
            return Err(ControllerError::InvalidRatios);
        }
        if !(0.0 < config.ewma_alpha && config.ewma_alpha <= 1.0) {
            return Err(ControllerError::InvalidAlpha);
        }
        if config.shrink_factor == 0 {
            return Err(ControllerError::InvalidShrinkFactor);
        }
        Ok(Self {
            config,
            state: ControllerState::default(),
        })
    }

    /// Create a controller with the Python parameter defaults.
    pub fn with_defaults() -> Self {
        Self::new(ControllerConfig::default()).expect("default controller config is valid")
    }

    /// Run one control cycle using an explicit monotonic timestamp.
    pub fn decide(&mut self, metrics: ControllerMetrics, now: f64) -> Decision {
        let ewma = self.update_ewma(metrics.queue_ratio);
        let in_cooldown = now - self.state.last_decision_at < self.config.cooldown_s;
        if in_cooldown {
            return Decision {
                action: Action::Hold,
                target_workers: metrics.worker_count,
                ewma_depth: ewma,
                in_cooldown: true,
                reason: "in cooldown".to_owned(),
            };
        }

        let in_band = self.config.low_ratio <= ewma && ewma <= self.config.high_ratio;
        if in_band {
            self.state.out_of_bounds_count = 0;
            return Decision {
                action: Action::Hold,
                target_workers: metrics.worker_count,
                ewma_depth: ewma,
                in_cooldown: false,
                reason: "within target band".to_owned(),
            };
        }

        self.state.out_of_bounds_count = self.state.out_of_bounds_count.saturating_add(1);
        if self.state.out_of_bounds_count < self.config.hysteresis_samples {
            return Decision {
                action: Action::Hold,
                target_workers: metrics.worker_count,
                ewma_depth: ewma,
                in_cooldown: false,
                reason: format!(
                    "hysteresis ({}/{})",
                    self.state.out_of_bounds_count, self.config.hysteresis_samples
                ),
            };
        }

        let mut target = metrics.worker_count;
        let mut action = Action::Hold;
        let mut reason = String::new();
        if ewma > self.config.high_ratio {
            let is_slow = metrics.task_elapsed > self.slow_task_seconds()
                && metrics.queue_ratio > self.config.high_ratio;
            let step = if is_slow {
                action = Action::GrowFast;
                self.config.grow_step.saturating_mul(2)
            } else {
                action = Action::Grow;
                self.config.grow_step
            };
            target = metrics
                .worker_count
                .saturating_add(step)
                .min(metrics.worker_max);
            reason = if is_slow {
                format!("slow tasks detected, grow fast +{step}")
            } else {
                format!(
                    "queue ratio {:.3} > {}, grow +{}",
                    ewma, self.config.high_ratio, step
                )
            };
        } else if ewma < self.config.low_ratio {
            target = (metrics.worker_count / self.config.shrink_factor).max(metrics.worker_min);
            action = Action::Shrink;
            reason = format!(
                "queue ratio {:.3} < {}, shrink to {}",
                ewma, self.config.low_ratio, target
            );
        }

        self.state.last_decision_at = now;
        self.state.out_of_bounds_count = 0;
        self.state.decisions_total = self.state.decisions_total.saturating_add(1);
        Decision {
            action,
            target_workers: target,
            ewma_depth: ewma,
            in_cooldown: false,
            reason,
        }
    }

    /// Reset EWMA, hysteresis, cooldown, and decision counters.
    pub fn reset(&mut self) {
        self.state = ControllerState::default();
    }

    /// Return an observable state snapshot using a caller-provided clock.
    pub fn state(&self, now: f64) -> StateSnapshot {
        StateSnapshot {
            ewma: round(self.state.ewma, 4),
            out_of_bounds_count: self.state.out_of_bounds_count,
            last_decision_elapsed: if self.state.last_decision_at != 0.0 {
                round(now - self.state.last_decision_at, 2)
            } else {
                0.0
            },
            decisions_total: self.state.decisions_total,
            low_ratio: self.config.low_ratio,
            high_ratio: self.config.high_ratio,
            ewma_alpha: self.config.ewma_alpha,
            hysteresis_samples: self.config.hysteresis_samples,
            cooldown_s: self.config.cooldown_s,
            grow_step: self.config.grow_step,
            shrink_factor: self.config.shrink_factor,
        }
    }

    /// Return the effective configuration.
    pub const fn config(&self) -> ControllerConfig {
        self.config
    }

    fn slow_task_seconds(&self) -> f64 {
        self.config.task_timeout * self.config.slow_task_ratio
    }

    fn update_ewma(&mut self, queue_ratio: f64) -> f64 {
        if self.state.decisions_total == 0 && self.state.ewma == 0.0 {
            self.state.ewma = queue_ratio;
        } else {
            self.state.ewma = self.config.ewma_alpha * queue_ratio
                + (1.0 - self.config.ewma_alpha) * self.state.ewma;
        }
        self.state.ewma
    }
}

impl Default for LoadAdaptiveController {
    fn default() -> Self {
        Self::with_defaults()
    }
}

fn round(value: f64, places: i32) -> f64 {
    let factor = 10_f64.powi(places);
    (value * factor).round() / factor
}

File: systems/python-reference-runtime/l1/kernel/load_adaptive.py
"""Load-adaptive thread pool controller.

Pure algorithm (no I/O, no OS dependencies) — the control law is a
self-contained numerical function suitable for FFI porting to Rust.

Architecture
------------
    Load signals (periodic sampling)     LoadAdaptiveController (pure)
    ──────────────────────                ──────────────────────────────
      queue_ratio     ──► EWMA smoothing ──► target interval [LOW, HIGH]
      completion_rate ──► hysteresis       ├─ below LOW  → shrink (×1/2)
      active_ratio    ──► cooldown timer   ├─ above HIGH → grow (+2)
      task_elapsed    ──► output target    └─ in band    → hold
                                           │
                                           ▼
                               ThreadPoolWorker.grow()/shrink()

Usage
-----
    controller = LoadAdaptiveController()
    action = controller.decide(metrics)   # returns Action enum
"""

from __future__ import annotations

import time
from dataclasses import dataclass
from enum import Enum, auto
from typing import Any

from l1.kernel.params.api import (
    LOAD_ADAPTIVE_COOLDOWN_S,
    LOAD_ADAPTIVE_EWMA_ALPHA,
    LOAD_ADAPTIVE_GROW_STEP,
    LOAD_ADAPTIVE_HIGH_RATIO,
    LOAD_ADAPTIVE_HYSTERESIS_SAMPLES,
    LOAD_ADAPTIVE_LOW_RATIO,
    LOAD_ADAPTIVE_SHRINK_FACTOR,
    LOAD_ADAPTIVE_SLOW_TASK_RATIO,
    WORKER_POOL_TASK_TIMEOUT,
)

logger = __import__("logging").getLogger(__name__)


class Action(Enum):
    """Controller decision emitted by decide()."""

    HOLD = auto()
    GROW = auto()
    SHRINK = auto()
    GROW_FAST = auto()


@dataclass
class ControllerMetrics:
    """Snapshot of thread-pool signals fed into decide()."""

    queue_ratio: float = 0.0
    completion_rate: float = 0.0
    active_ratio: float = 0.0
    task_elapsed: float = 0.0
    worker_count: int = 0
    worker_min: int = 1
    worker_max: int = 32


@dataclass
class ControllerState:
    """Persistent state of the controller between sampling cycles."""

    ewma: float = 0.0
    _out_of_bounds_count: int = 0
    _last_decision_at: float = 0.0
    decisions_total: int = 0


@dataclass
class Decision:
    """Result of a single decide() call."""

    action: Action = Action.HOLD
    target_workers: int = 0
    ewma_depth: float = 0.0
    in_cooldown: bool = False
    reason: str = ""


class LoadAdaptiveController:
    """Load-adaptive controller for thread-pool sizing.

    This is a **pure algorithm** — no file I/O, no network, no OS calls.
    It consumes ControllerMetrics snapshots and returns Decision.
    The caller is responsible for periodic sampling and applying the action.
    """

    def __init__(
        self,
        low_ratio: float = LOAD_ADAPTIVE_LOW_RATIO,
        high_ratio: float = LOAD_ADAPTIVE_HIGH_RATIO,
        ewma_alpha: float = LOAD_ADAPTIVE_EWMA_ALPHA,
        hysteresis_samples: int = LOAD_ADAPTIVE_HYSTERESIS_SAMPLES,
        cooldown_s: float = LOAD_ADAPTIVE_COOLDOWN_S,
        grow_step: int = LOAD_ADAPTIVE_GROW_STEP,
        shrink_factor: int = LOAD_ADAPTIVE_SHRINK_FACTOR,
        slow_task_ratio: float = LOAD_ADAPTIVE_SLOW_TASK_RATIO,
        task_timeout: float = WORKER_POOL_TASK_TIMEOUT,
    ) -> None:
        if not (0.0 <= low_ratio < high_ratio <= 1.0):
            raise ValueError(f"need 0 <= low_ratio ({low_ratio}) < high_ratio ({high_ratio}) <= 1.0")
        if not (0.0 < ewma_alpha <= 1.0):
            raise ValueError(f"ewma_alpha must be in (0, 1], got {ewma_alpha}")
        self._low_ratio = low_ratio
        self._high_ratio = high_ratio
        self._ewma_alpha = ewma_alpha
        self._hysteresis_samples = hysteresis_samples
        self._cooldown_s = cooldown_s
        self._grow_step = grow_step
        self._shrink_factor = shrink_factor
        self._slow_task_ms = task_timeout * slow_task_ratio
        self._state = ControllerState()

    def decide(self, metrics: ControllerMetrics, now: float | None = None) -> Decision:
        """Run one control-law cycle and return a decision."""
        if now is None:
            now = time.monotonic()

        ewma = self._update_ewma(metrics.queue_ratio)

        in_cooldown = (now - self._state._last_decision_at) < self._cooldown_s
        if in_cooldown:
            return Decision(
                action=Action.HOLD,
                target_workers=metrics.worker_count,
                ewma_depth=ewma,
                in_cooldown=True,
                reason="in cooldown",
            )

        in_band = self._low_ratio <= ewma <= self._high_ratio
        if in_band:
            self._state._out_of_bounds_count = 0
            return Decision(
                action=Action.HOLD,
                target_workers=metrics.worker_count,
                ewma_depth=ewma,
                in_cooldown=False,
                reason="within target band",
            )

        self._state._out_of_bounds_count += 1
        if self._state._out_of_bounds_count < self._hysteresis_samples:
            return Decision(
                action=Action.HOLD,
                target_workers=metrics.worker_count,
                ewma_depth=ewma,
                in_cooldown=False,
                reason=f"hysteresis ({self._state._out_of_bounds_count}/{self._hysteresis_samples})",
            )

        target = metrics.worker_count
        action: Action = Action.HOLD
        reason = ""

        if ewma > self._high_ratio:
            is_slow = metrics.task_elapsed > self._slow_task_ms and metrics.queue_ratio > self._high_ratio
            if is_slow:
                step = self._grow_step * 2
                action = Action.GROW_FAST
                reason = f"slow tasks detected, grow fast +{step}"
            else:
                step = self._grow_step
                action = Action.GROW
                reason = f"queue ratio {ewma:.3f} > {self._high_ratio}, grow +{step}"
            target = min(metrics.worker_count + step, metrics.worker_max)

        elif ewma < self._low_ratio:
            shrunk = max(metrics.worker_count // self._shrink_factor, metrics.worker_min)
            target = shrunk
            action = Action.SHRINK
            reason = f"queue ratio {ewma:.3f} < {self._low_ratio}, shrink to {target}"

        self._state._last_decision_at = now
        self._state._out_of_bounds_count = 0
        self._state.decisions_total += 1

        return Decision(
            action=action,
            target_workers=target,
            ewma_depth=ewma,
            in_cooldown=False,
            reason=reason,
        )

    def reset(self) -> None:
        """Reset the controller to its initial state."""
        self._state = ControllerState()

    def state(self) -> dict[str, Any]:
        """Return serialisable controller state for observability."""
        return {
            "ewma": round(self._state.ewma, 4),
            "out_of_bounds_count": self._state._out_of_bounds_count,
            "last_decision_elapsed": (
                round(time.monotonic() - self._state._last_decision_at, 2) if self._state._last_decision_at else 0.0
            ),
            "decisions_total": self._state.decisions_total,
            "low_ratio": self._low_ratio,
            "high_ratio": self._high_ratio,
            "ewma_alpha": self._ewma_alpha,
            "hysteresis_samples": self._hysteresis_samples,
            "cooldown_s": self._cooldown_s,
            "grow_step": self._grow_step,
            "shrink_factor": self._shrink_factor,
        }

    def _update_ewma(self, queue_ratio: float) -> float:
        """Update and return the smoothed queue-ratio signal."""
        if self._state.decisions_total == 0 and self._state.ewma == 0.0:
            self._state.ewma = queue_ratio
        else:
            self._state.ewma = self._ewma_alpha * queue_ratio + (1.0 - self._ewma_alpha) * self._state.ewma
        return self._state.ewma
